src/controllers/__tests__/authenticationController.test.ts

```diff
@@ -0,0 +1,367 @@
+Gmg7yDYVm/kETj3JQ/drUzKIbj6t9LXvUizOUa2VSMJ0yZTYsnDHuywi8nf0uhgO
+5pAca0aJLJ792hEByOx+EeUSN3C3i35vfbn8gwYoAHjrVA8mJrAEsawRbdVpNj6j
+IWAKTmsZK0YLdcNzWshSYW9wkJNykeXHUgKk2YzGUIacMgC+fF3v3xL82xk+eLez
+dP5wlrzkPz8JKHMIomF5j/VLuggSZx0XdQRvZrkeQUbJqRy2iXa43B+OlEiNvd2Q
+0AiXur/MhvID+Ni/hBIoeIDyvvoBoiCTZbVvyRnBds8BAoGBDIfqHTTcfXlzfipt
+4+idRuuzzhIXzQOcB0+8ywqmgtE4g9EykC7fEcynOavv08rOSNSbYhjLX24xUSrd
+19lckZIvH5U9nJxnyfwrMGrorCA2uPtck8N9sTB5UWif31w/eDVMv30jRUyMel7l
+tp96zxcPThT1O3N4zM2Otk5q2DvFAoGBDBngF4G9dJ5a511dyYco2agMJTvJLIqn
+kKw24KOTqZ5BZKyIea4yCy9cRmNN84ccOy3jBuzSFLNJMYqdDCzH46u0I4anft83
+aqnVa4jOwjZXoV9JCdFh3zKJUgPU4CW0MaTb30n3U4BAOgkHzRFt55tGT6xRU1J+
+jX5s03BFfQ/pAoGBCsRqtUfrweEvDRT2MeR56Cu153cCfoYAdwPcDHeNVlDih9mk
+4eF0ib3ZXyPPQqQ8FrahAWyeq9Rqif0UfFloQiljVncNZtm6EQQeNE9YuDZB7zcF
+eG59PViSlhIZdXq1itv5o3yqZux8tNV/+ykUBIgi/YvioH/7J7flTd8Zzc2lAoGB
+CqdVNRzSETPBUGRQx7Yo7scWOkmSaZaAw8v6XHdm7zQW2m1Tkd0czeAaWxXecQKI
+hkl10Ij6w6K8U9N3RFrAeN6YL5bDK92VSmDPNmcxsKZrK/VZtj0S74/sebpJ1jUb
+mYFM2h6ikm8dHHsK1S39FqULl+VbjAHazPN7GAOGCf7RAoGBAc0J9j+MHYm4M18K
+AW2UB26qvdc8PSXE6i4YQAsg2RBgtf6u4jVAQ8a8OA4vnGG9VIrR0bD/hFTfczg/
+ZbWGZ+42VH2eDGouiadR4rWzJNQKjWd98Y5PzxdEjd6nneJ68iNqGbKOT6jXu8qj
+nCnxP/vK5rgVHU3nQfq+e/B6FVWZ
+-----END PRIVATE KEY-----
+`;
+const samlCallbackUrl = "http://italia-backend/assertionConsumerService";
+const samlIssuer = "http://italia-backend";
+const samlAttributeConsumingServiceIndex = 0;
+
+// user constant
+const aTimestamp = 1518010929530;
+
+const aFiscalNumber = "GRBGPP87L04L741X" as FiscalCode;
+const anEmailAddress = "garibaldi@example.com" as EmailAddress;
+
+// authentication constant
+const mockToken =
+  "c77de47586c841adbd1a1caeb90dce25dcecebed620488a4f932a6280b10ee99a77b6c494a8a6e6884ccbeb6d3fe736b";
+
+// mock for a valid User
+const mockedUser: User = {
+  created_at: aTimestamp,
+  family_name: "Garibaldi",
+  fiscal_code: aFiscalNumber,
+  name: "Giuseppe Maria",
+  nameID: "garibaldi",
+  nameIDFormat: "urn:oasis:names:tc:SAML:2.0:nameid-format:transient",
+  preferred_email: anEmailAddress,
+  sessionIndex: "123sessionIndex",
+  spid_idp: "xxx",
+  token: mockToken
+};
+
+// validUser has all every field correctly set.
+const validUserPayload = {
+  email: anEmailAddress,
+  familyName: "Garibaldi",
+  fiscalNumber: aFiscalNumber,
+  issuer: {
+    _: "xxx"
+  },
+  name: "Giuseppe Maria",
+  nameID: "garibaldi",
+  nameIDFormat: "urn:oasis:names:tc:SAML:2.0:nameid-format:transient",
+  sessionIndex: "123sessionIndex"
+};
+// invalidUser lacks the required email field.
+const invalidUserPayload = {
+  familyName: "Garibaldi",
+  fiscalNumber: aFiscalNumber,
+  issuer: {
+    _: "xxx"
+  },
+  name: "Giuseppe Maria",
+  nameID: "garibaldi",
+  nameIDFormat: "urn:oasis:names:tc:SAML:2.0:nameid-format:transient",
+  sessionIndex: "123sessionIndex"
+};
+
+const mockSet = jest.fn();
+const mockGet = jest.fn();
+const mockDel = jest.fn();
+jest.mock("../../services/redisSessionStorage", () => {
```

Comment:
most of the times you can just call 
```
jest.mock("../../services/redisSessionStorage")
import * as redis from "../../services/redisSessionStorage"
```
then use `(redis.foo as jest.Mock).mockImplementation() ...` when you need them